### PR TITLE
Implement query() to give possible schedule times

### PR DESCRIPTION
**Pull request by yicyao**
Finished this task:
Imagine you're working on the calendar team and are responsible for adding the "find a meeting" feature. Using the existing API, you'll need to implement a feature `query()` that given the meeting information, it will return the times when the meeting could happen that day.

`query()` passes all 21 tests in FindMeetingQueryTests.java
xxAdiGBNbNj

'If you have time left over this week and you're looking for a challenge: What's the algorithmic complexity of your algorithm?'
Time complexity: O(mlogm) + O(mn) + O(m) + O(m) --> O(mn) where m = size of `Collection<Event>` events, n = size of `request.getAttendees()`
`Collections.sort(): `O(mlogm)
`getBusyTimes():` O(mn), 2 for loops to loop through all events and attendees
`mergeIntervals():` O(m). goes through all events to merge intervals
`getAvailableTimes()`: O(m): goes through all event times to find available times

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,80 @@
...
+  // gets available times of correct duration
+  private List<TimeRange> getAvailableTimes(List<TimeRange> busyTimes, long duration) {
+    List<TimeRange> availableTimes = new ArrayList<>();
+
+    // finds gaps of requested duration time
+    int currentStart = TimeRange.START_OF_DAY;
+    for (TimeRange time : busyTimes) {
+      if (time.start() - currentStart >= duration) {
+        availableTimes.add(
+            TimeRange.fromStartEnd(currentStart, time.start(), /* inclusive=*/false));
```
Comment (nikhitashanker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: space between / and true

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,80 @@
...
+      }
+      currentStart = time.end();
+    }
+
+    // finds gap of requested duration time between last meeting and end of day
+    if (TimeRange.END_OF_DAY - currentStart >= duration) {
+      availableTimes.add(
+          TimeRange.fromStartEnd(currentStart, TimeRange.END_OF_DAY, /* inclusive=*/true));
```
Comment (nikhitashanker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: ditto

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,80 @@
...
+    return busyTimes;
+  }
+
+  // merges intervals of overlapping events
+  private List<TimeRange> mergeIntervals(List<TimeRange> intervals) {
+    if (intervals == null || intervals.size() <= 1) {
+      return intervals;
+    }
+
+    List<TimeRange> result = new ArrayList<>();
```
Comment (nikhitashanker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: Up to personal preference but maybe rename this to something more descriptive than result.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,80 @@
...
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> busyTimes = getBusyTimes(events, request);
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    busyTimes = mergeIntervals(busyTimes);
+    return getAvailableTimes(busyTimes, request.getDuration());
+  }
+
+  // gets busy times for all request attendees
+  private ArrayList<TimeRange> getBusyTimes(Collection<Event> events, MeetingRequest request) {
```
Comment (stephanie-fu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It's probably good to return a type as generic as possible (just like your other private methods returning List) unless you actually need ArrayList-specific functionality

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,80 @@
...
+    ArrayList<TimeRange> busyTimes = new ArrayList<>();
```
Comment (stephanie-fu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Same here, declaring as List but instantiating ArrayList

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (yicyao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (yicyao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (yicyao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (yicyao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,80 @@
...
+          busyTimes.add(event.getWhen());
+        }
+      }
+    }
+    return busyTimes;
+  }
+
+  // merges intervals of overlapping events
+  private List<TimeRange> mergeIntervals(List<TimeRange> intervals) {
+    if (intervals == null || intervals.size() <= 1) {
```
Comment (hlz03) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: use intervals.isEmpty()

**Review by hlz03 — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (yicyao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

**Pull request merged**